pep-9999.rst

```diff
@@ -0,0 +1,1037 @@
+PEP: 9999
+Title: Graceful cooperation between external and Python package managers
+Author: Geoffrey Thomas <geofft@ldpreload.com>, Matthias Klose <doko@ubuntu.com>, Filipe Laíns <lains@riseup.net>, Donald Stufft <donald@python.org>, Tzu-Ping Chung <uranusjr@gmail.com>, Stefano Rivera <stefanor@debian.org>, Elana Hashman <ehashman@debian.org>, Pradyun Gedam <mail@pradyunsg.me>
+Discussions-To: TODO discourse
+Status: Draft
+Type: Informational
+Content-Type: text/x-rst
+Created: 2021-05-18
+Post-History:
+
+Abstract
+========
+
+This PEP proposes a mechanism to indicate that a certain
+Python interpreter's package installation is managed by
+external means (such as an OS package manager) and that
+Python-specific package management tools like pip should, by
+default, neither install nor remove packages in an
+interpreter-wide context, guiding the end user towards using
+a virtual environment.
+
+It also standardizes an interpretation of the ``sysconfig``
+schemes so that, if a Python-specific package manager is
+specifically instructed to install a package in an
+interpreter-wide context, it can do so in a manner that
+avoids conflicting with the external package manager and
+reduces the risk of breaking software shipped by the
+external package manager.
+
+Motivation
+==========
+
+Thanks to Python's immense popularity, software
+distributions (by which we mean Linux and other OS
+distributions as well as overlay distributions like Homebrew
+and MacPorts) generally ship Python for two purposes: as a
+software package to be used in its own right by end users,
+and as an implementation language for other software in the
+distribution.
+
+For example, many distributions ship a ``/usr/bin/python3``
+command which is both the ``python3`` command available to
+end users as well as the ``#!/usr/bin/python3`` shebang for
+Python-language software. Because there are no official
+binary releases of Python for Linux/UNIX, almost all Python
+end users on these OSes use the Python interpreter built and
+shipped with their distribution.
+
+These are typically the same command, which means that if an
+end user installs a Python package using a tool like ``pip``
+outside the context of a virtual environment, the new
+package is visible to Python-language software shipped by
+the distribution. If the new package is not
+backwards-compatible, it may break software shipped by the
```

Comment:
I'm not sure what "backwards-compatible" means here. More importantly, the process by which that new package was installed may break the distribution when it resolves and updates dependencies.

Reply:
I mean regular old API backwards compatibility - in particular, the `python-requests` in Debian 7 (some 0.x) had `.json` as a property but requests 1.x just uses `.json()` as a normal function, so if you `sudo pip install -U requests`, you break code written against the OS version of requests.

I think you're referencing another and also interesting problem about breaking the distribution... can we get into a state where `pip` leaves dpkg unable to deal with what's on the filesystem? If so, let's mention/document it. (I was thinking about the case where Python-language code such as `dnf` breaks.)

Reply:
Oh I realized there was no antecedent to _what_ the new package is supposed to be ackwwards-compatible with. I expanded this sentence, it should make more sense now.